src/dune_rules/pkg_rules.ml

```diff
@@ -621,6 +623,109 @@ module Substitute = struct
 end
 
 module Run_with_path = struct
+  module Output : sig
+    type t
+
+    val io : t -> Process.Io.output Process.Io.t
+
+    val v
+      :  ?accepted_exit_codes:int Predicate.t
+      -> ?pkg_name:Dune_lang.Package_name.t
+      -> string
+      -> t
+
+    val consume_and_print_error : t -> Display.t -> code:int -> loc:Loc.t -> unit
+  end = struct
+    type state =
+      | Open
+      | Closed
+
+    type t =
+      { pkg_name : Dune_pkg.Package_name.t option
+      ; fn : Dpath.t
+      ; io : Process.Io.output Process.Io.t
+      ; accepted_exit_codes : int Predicate.t
+      ; mutable state : state
+      }
+
+    let prefix = "dune-pkg"
+    let exit_code_zero = Predicate.create (fun x -> x = 0)
+    let has_output error = error <> ""
+
+    let v ?(accepted_exit_codes = exit_code_zero) ?pkg_name suffix =
+      let fn = Temp.create File ~prefix ~suffix in
```

Comment:
I think we can improve the resource management here:
- `v` kinds of sound like a constructor (we tend to use the `create` name for this), but it's actually opening a file.
- if you can restructure that to a "context manager" this will be easier to reason about. something like `val with_ : ?accepted_exit_codes -> ?pkg_name -> string -> f:(t -> 'a) -> 'a`, a bit like `In_channel.with_open`. that way you don't have to deal with `state`, `close` etc.

Reply:
- Yes, the `v` is an old `Irmin` habits... 
- It would make the code much simpler! Thanks